torchgeo/datasets/geo.py

```diff
@@ -676,46 +759,70 @@ def _load_image(self, index: int) -> Tuple[Tensor, Tensor]:
         return tensor, label
 
 
-class ZipDataset(GeoDataset):
-    """Dataset for merging two or more GeoDatasets.
+class IntersectionDataset(GeoDataset):
+    """Dataset representing the intersection of two GeoDatasets.
+
+    This allows users to do things like:
+
+    * Combine image and target labels and sample from both simultaneously
+      (e.g. Landsat and CDL)
+    * Combine datasets for multiple image sources for multimodal learning or data fusion
+      (e.g. Landsat and Sentinel)
+
+    These combinations require that all queries are present in *both* datasets,
+    and can be combined using an :class:`IntersectionDataset`:
+
+    .. code-block:: python
 
-    For example, this allows you to combine an image source like Landsat8 with a target
-    label like CDL.
+       dataset = landsat & cdl
     """
 
-    def __init__(self, datasets: Sequence[GeoDataset]) -> None:
+    def __init__(
+        self,
+        dataset1: GeoDataset,
+        dataset2: GeoDataset,
+        collate_fn: Callable[
+            [Sequence[Dict[str, Any]]], Dict[str, Any]
+        ] = concat_samples,
+    ) -> None:
         """Initialize a new Dataset instance.
 
         Args:
-            datasets: list of datasets to merge
+            dataset1: the first dataset
+            dataset2: the second dataset
+            collate_fn: function used to collate samples
 
         Raises:
-            ValueError: if datasets contains non-GeoDatasets, do not overlap, are not in
-                the same :term:`coordinate reference system (CRS)`, or do not have the
-                same resolution
+            ValueError: if either dataset is not a :class:`GeoDataset`
         """
-        for ds in datasets:
+        super().__init__()
+        self.datasets = [dataset1, dataset2]
+        self.collate_fn = collate_fn
+
+        for ds in self.datasets:
             if not isinstance(ds, GeoDataset):
-                raise ValueError("ZipDataset only supports GeoDatasets")
-
-        crs = datasets[0].crs
-        res = datasets[0].res
-        for ds in datasets:
-            if ds.crs != crs:
-                raise ValueError("Datasets must be in the same CRS")
-            if not math.isclose(ds.res, res):
-                # TODO: relax this constraint someday
-                raise ValueError("Datasets must have the same resolution")
-
-        self.datasets = datasets
-        self.crs = crs
-        self.res = res
+                raise ValueError("IntersectionDataset only supports GeoDatasets")
+
+        self._crs = dataset1.crs
+        self.res = dataset1.res
 
-        # Make sure datasets have overlap
-        try:
-            self.bounds
-        except ValueError:
-            raise ValueError("Datasets have no overlap")
+        # Force dataset2 to have the same CRS/res as dataset1
+        dataset2.crs = dataset1.crs
```

Comment:
Why is this necessary? (this is present in both Union and Intersection)

Reply:
If the two datasets have a different CRS/res, we need to ensure that they have a matching CRS/res. Otherwise, the combined R-tree index would be meaningless. I added a getter/setter to GeoDataset that updates the entire index when you try to set the CRS to a different CRS.

Reply:
Makes sense. Should we warn users that if they use a dataset in an Intersection/Union then it might have its properties changed?

Reply:
Done. I ended up using a print statement instead of a warning since warnings aren't displayed by default.